webroot/panel/new_group.php

```diff
@@ -0,0 +1,172 @@
+    } else {
+        $group_name = $_POST['group_name'];
+    }
+
+    $group_uid = $group_type_prefix . $group_name;
+
+    $new_group = new UnityGroup($group_uid, $LDAP, $SQL, $MAILER, $REDIS, $WEBHOOK);
+    $new_group->requestGroup(
+        $USER->getUID(),
+        $group_type_slug,
+        $group_name,
+        $SEND_PIMESG_TO_ADMINS,
+        $group_start_date,
+        $group_end_date
+    );
+    header("Refresh:0");
+}
+
+?>
+
+<h1>Request New Group</h1>
+<hr>
+
+<?php
+if (count($pending_requests) > 0) {
+    echo "<h5>Pending Requests</h5>";
+    echo "<table>";
+    foreach ($pending_requests as $request) {
+        $requested_owner = $USER;
+        echo "<tr class='pending_request'>";
+        echo "<td> 
+        <div class='type' style='border-radius: 5px; padding-left: 10px; color: white; 
+        padding-right: 10px; text-align: center; font-size: 12px; 
+        background-color: #800000'>" . $USER->getTypeNameFromSlug($types, $request['group_type']) . "</div></td>";
+        echo "<td>" . $request['group_name'] . "</td>";
+        echo "<td>" . date("jS F, Y", strtotime($request['requested_on'])) . "</td>";
+        echo "<td></td>";
+        echo "</tr>";
+    }
+    echo "</table>";
+
+    echo "<hr>";
+}
+?>
+
+<form id="newGroupForm" action="" method="POST">
+    <p>Fill in the following information to request a new group</p>
+    <div>
+        <strong>Type of Group</strong><br>
+        <?php
+        foreach ($types as $type) {
+            if ($type['slug'] == 'pi') {
+                echo "<label><input type='radio' name='group_type' 
+                value='" . $type["prefix"] . "-" . $type["slug"] . "-" .
+                $type['time_limited'] . "-" . $type['isNameable'] . "' checked> " .
+                $type["name"] . "</label><br>";
+            } else {
+                echo "<label><input type='radio' name='group_type' 
+                value='" . $type["prefix"] . "-" . $type["slug"] . "-" .
+                $type['time_limited'] . "-" . $type['isNameable'] . "'> "
+                . $type["name"] . "</label><br>";
+            }
+        }
+        ?>
+        <div id="dateSelector" style="display: none; margin-top: 15px">
+            <strong>Select start and end dates</strong><br>
+            <label>Start Date: &nbsp;&nbsp;<input type='date' name='group_start_date'></label><br>
+            <label>End Date: &nbsp;&nbsp;<input type='date' name='group_end_date'></label><br>
+        </div>
+        <div id="nameInputBox" style="margin-top: 10px;">
+            <strong>Name (cannot have spaces)&nbsp;&nbsp;</strong><br>
+            <input type="text" name="group_name" placeholder="name_of_the_group" style="margin-bottom: 15px"><br>
+            <div style="color: red; font-size: 0.8rem; display: none; margin-top: -10px;" 
+            id="groupNameError">Error Occured<br></div>
+        </div>
+    </div>
+    <input style='margin-top: 10px;' type='submit' value='Request Group' id="requestGroupButton">
```

Comment:
If the user is already a PI, the system allows the user to request a PI group, but the request doesn't go through, giving no feedback to the user. We didn't think about this case when we initially talked about the generalization.

To resolve this, let's add a new column to the `groupTypes` table with the field name `exclusiveOwner`. If this value is `1`, then then each user can **own** only 1 kind of this group. For example, this would be 1 for the PI group, and 0 for the others, because a PI would only have 1 PI group assigned to itself, but they can still be a member of other PI groups, they just can't be owners. Let me know if that is confusing and I can try to explain it better.